contracts/utils/Checkpoints.sol

```diff
@@ -15,21 +16,233 @@ import "./math/SafeCast.sol";
  * _Available since v4.5._
  */
 library Checkpoints {
-    struct Checkpoint {
-        uint32 _blockNumber;
+    struct Checkpoint224 {
+        uint32 _key;
         uint224 _value;
     }
 
+    function latest(Checkpoint224[] storage self) internal view returns (uint224) {
+        uint256 pos = self.length;
+        return pos == 0 ? 0 : _unsafeAccess(self, pos - 1)._value;
+    }
+
+    function push(
+        Checkpoint224[] storage self,
+        uint32 key,
+        uint224 value
+    ) internal returns (uint224, uint224) {
+        uint256 pos = self.length;
+
+        if (pos > 0) {
+            // Use of memory is important here.
+            Checkpoint224 memory last = _unsafeAccess(self, pos - 1);
+
+            // Checkpoints keys must be increassing.
+            require(last._key <= key, "Checkpoint: invalid key");
+
+            // Update or push new checkpoint
+            if (last._key == key) {
+                _unsafeAccess(self, pos - 1)._value = value;
+            } else {
+                self.push(Checkpoint224({_key: key, _value: value}));
+            }
+            return (last._value, value);
+        } else {
+            self.push(Checkpoint224({_key: key, _value: value}));
+            return (0, value);
+        }
+    }
+
+    function lowerLookup(Checkpoint224[] storage self, uint32 key) internal view returns (uint224) {
+        uint256 length = self.length;
+        uint256 pos = _lowerDichotomicLookup(self, key, 0, length);
+        return pos == length ? 0 : _unsafeAccess(self, pos)._value;
+    }
+
+    function upperLookup(Checkpoint224[] storage self, uint32 key) internal view returns (uint224) {
+        uint256 length = self.length;
+        uint256 pos = _upperDichotomicLookup(self, key, 0, length);
+        return pos == 0 ? 0 : _unsafeAccess(self, pos - 1)._value;
+    }
+
+    function upperLookupRecent(Checkpoint224[] storage self, uint32 key) internal view returns (uint224) {
+        uint256 length = self.length;
+        uint256 offset = 1;
+
+        while (offset <= length && _unsafeAccess(self, length - offset)._key > key) {
+            offset <<= 1;
+        }
+
+        uint256 low = offset < length ? length - offset : 0;
+        uint256 high = length - (offset >> 1);
+        uint256 pos = _upperDichotomicLookup(self, key, low, high);
+
+        return pos == 0 ? 0 : _unsafeAccess(self, pos - 1)._value;
+    }
+
+    function _upperDichotomicLookup(
```

Comment:
I wasn't familiar with the term "dichotomic search/lookup", looking at the Wikipedia page it says binary search is a kind of dichotomic search, so this is a dichotomic search but also a binary search (right?). I would prefer to use the term "binary search" or "binary lookup" which is more widely known.

Reply:
AFAIK, dichotomic search and binary search are mostly equivalent. I believe that dichotomic is technically more precise, but if devs in our space are more familiar with the "binary search", I'm ok with using that instead